### PR TITLE
Issues entity

### PR DESCRIPTION
**Pull request by liuxess**

### CODE REVIEW

File: src/main/java/com/kustudents/issuetracker/controller/viwIssuesController.java [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+
+
+import java.util.List;
+
+import org.springframework.web.bind.annotation.*;
+import com.kustudents.issuetracker.data.enitity.viwIssues;
+import com.kustudents.issuetracker.service.viwIssuesService;
+
+@RestController
+@RequestMapping("/api/Issues")
```
Comment (ValdPiek) [inline src/main/java/com/kustudents/issuetracker/controller/viwIssuesController.java:11]: I think, there is a way to not include "/api/" in the mapping (at least we didn't have to in my last project). I will look into that

File: src/main/java/com/kustudents/issuetracker/controller/viwIssuesController.java [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+public class viwIssuesController {
```
Comment (ValdasPiekus) [inline src/main/java/com/kustudents/issuetracker/controller/viwIssuesController.java]: I have noticed, that you used "viw" or "v" at the beginnings of classes or objects. In my opinion, this information is unnecessary while creating code of the API, and also it reduces readability - when we will have 4 more controllers, models, repositories, and services it will be hard to follow the code. My suggestion: drop the "viw" and "v"

File: src/main/java/com/kustudents/issuetracker/controller/viwIssuesController.java [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+    private final viwIssuesService vIssuesService;
+    
+    public viwIssuesController(viwIssuesService vIssuesService){
+        this.vIssuesService = vIssuesService;
+    }
```
Comment (ValdasPiekus) [inline src/main/java/com/kustudents/issuetracker/controller/viwIssuesController.java]: "viw" and "v".

We can use the annotation "@RequiredArgsConstructor" (should be placed next to RestController and @RestMapping annotations), which will do the job of creating the class constructor, thus, we will not need lines 15-17.

File: src/main/java/com/kustudents/issuetracker/controller/viwIssuesController.java [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+
+    @GetMapping("/")
+    public List<viwIssues> getAll() {
+        return vIssuesService.getAllIssues();
+    }
+
+    @GetMapping("/{id}")
+    public viwIssues getPatient(@PathVariable("id") Long id) {
```
Comment (ValdasPiekus) [inline src/main/java/com/kustudents/issuetracker/controller/viwIssuesController.java]: "getPatient" to "getIssue"

File: src/main/java/com/kustudents/issuetracker/controller/viwIssuesController.java (hunk 2 above)
Comment (ValdasPiekus) [reply src/main/java/com/kustudents/issuetracker/controller/viwIssuesController.java]: We also need to discuss if we want to use singular or plural naming of classes (IssueController vs IssuesControler - IssueService vs IssuesService)

File: src/main/java/com/kustudents/issuetracker/data/enitity/viwIssues.java [hunk 5]
```diff
@@ -0,0 +1,28 @@
+package com.kustudents.issuetracker.data.enitity;
+
+
+import javax.persistence.*;
+import java.time.LocalDateTime;
+import lombok.Data;
+
+@Entity(name = "viw_Issues")
+@Data
+public class viwIssues {
```
Comment (ValdasPiekus) [inline src/main/java/com/kustudents/issuetracker/data/enitity/viwIssues.java]: "viw"

And changing to "Issue" would be better. Here we define what structure a single issue object has

File: src/main/java/com/kustudents/issuetracker/data/enitity/viwIssues.java [hunk 6]
```diff
@@ -0,0 +1,28 @@
...
+    
+    @Column
+    @Id
+    private int id;
+
+    @Column
+    private LocalDateTime closed;
+
+    @Column
+    private String name;
```
Comment (ValdasPiekus) [inline src/main/java/com/kustudents/issuetracker/data/enitity/viwIssues.java]: "@NotBlank(message = "Enter the name of the Issue")" can be added for additional security

File: src/main/java/com/kustudents/issuetracker/data/enitity/viwIssues.java [hunk 7]
```diff
@@ -0,0 +1,28 @@
...
+
+    @Column
+    private String description;
+
+    @Column
+    private int importance;
```
Comment (ValdasPiekus) [inline src/main/java/com/kustudents/issuetracker/data/enitity/viwIssues.java]: "@NotNull()" can be added for additional security

File: src/main/java/com/kustudents/issuetracker/data/repo/viwIssuesRepo.java [hunk 8]
```diff
@@ -0,0 +1,9 @@
+package com.kustudents.issuetracker.data.repo;
+
+import com.kustudents.issuetracker.data.enitity.viwIssues;
+
+import org.springframework.data.jpa.repository.JpaRepository;
+
+public interface viwIssuesRepo extends JpaRepository<viwIssues, Long>{
```
Comment (ValdasPiekus) [inline src/main/java/com/kustudents/issuetracker/data/repo/viwIssuesRepo.java]: "viw"

File: src/main/java/com/kustudents/issuetracker/service/viwIssuesService.java [hunk 9]
```diff
@@ -0,0 +1,35 @@
...
+import java.util.List;
+
+import com.kustudents.issuetracker.data.enitity.viwIssues;
+import com.kustudents.issuetracker.data.repo.viwIssuesRepo;
+
+import org.springframework.data.jpa.repository.JpaRepository;
+import org.springframework.stereotype.Service;
+
+@Service
+public class viwIssuesService{
```
Comment (ValdasPiekus) [inline src/main/java/com/kustudents/issuetracker/service/viwIssuesService.java:14]: "viw"

File: src/main/java/com/kustudents/issuetracker/service/viwIssuesService.java [hunk 10]
```diff
@@ -0,0 +1,35 @@
...
+    private final viwIssuesRepo vRepo;
```
Comment (ValdasPiekus) [inline src/main/java/com/kustudents/issuetracker/service/viwIssuesService.java:15]: "vRepo" to "issuesRepo", because if we add more repositories, we will have a mess

File: src/main/java/com/kustudents/issuetracker/service/viwIssuesService.java [hunk 11]
```diff
@@ -0,0 +1,35 @@
...
+
+    public viwIssuesService(viwIssuesRepo vRepo){
+        this.vRepo = vRepo;
+    }
```
Comment (ValdasPiekus) [inline src/main/java/com/kustudents/issuetracker/service/viwIssuesService.java]: Same as in controller, we can add "@RequiredArgsConstructor"

File: src/main/java/com/kustudents/issuetracker/service/viwIssuesService.java [hunk 12]
```diff
@@ -0,0 +1,35 @@
...
+
+    public viwIssues getIssueByID(Long id){
+        return vRepo.findById(id).orElseThrow(() -> new RuntimeException(""));
+    }
+
+    public List<viwIssues> getAllIssues(){
+        return vRepo.findAll();
+    }
+
+    public viwIssues createIssues(viwIssues vIssues){
```
Comment (ValdasPiekus) [inline src/main/java/com/kustudents/issuetracker/service/viwIssuesService.java:30]: So are we creating multiple issues here or are we creating just one? That's why we should use singular names

File: src/main/java/com/kustudents/issuetracker/service/viwIssuesService.java [hunk 13]
```diff
@@ -0,0 +1,35 @@
...
+        return vRepo.save(vIssues);
+    }
+
+    public void deleteById(Long id) {
```
Comment (ValdasPiekus) [inline src/main/java/com/kustudents/issuetracker/service/viwIssuesService.java:36]: if we have getIssueById, then we should have deleteIssueById - increased readability

File: src/main/java/com/kustudents/issuetracker/service/viwIssuesService.java [hunk 14]
```diff
@@ -0,0 +1,35 @@
...
+        vRepo.deleteById(id);
```
Comment (ValdasPiekus) [inline src/main/java/com/kustudents/issuetracker/service/viwIssuesService.java:37]: What happens, if an issue with the provided id does not exist?

**Comment by ValdPiek** [comment]
image
Also, we need to change the structure of packages to the one shown above. Repo should be named fully as repository, we can name the classes with "Repo"

File: src/main/java/com/kustudents/issuetracker/service/viwIssuesStepsService.java [hunk 15]
```diff
@@ -0,0 +1,46 @@
...
+import com.kustudents.issuetracker.model.entity.viwIssues;
+import com.kustudents.issuetracker.model.entity.viwIssuesSteps;
+import com.kustudents.issuetracker.repository.viwIssuesStepsRepo;
+import com.kustudents.issuetracker.service.factories.IssuesStepsFactory;
+
+import org.springframework.data.jpa.repository.JpaRepository;
+import org.springframework.stereotype.Service;
+
+@Service
+public class viwIssuesStepsService{
```
Comment (ValdasPiekus) [inline src/main/java/com/kustudents/issuetracker/service/viwIssuesStepsService.java:14]: Add @RequiredArgsConstructor, lines 19-22 will not be needed then

File: src/main/java/com/kustudents/issuetracker/service/viwIssuesStepsService.java [hunk 16]
```diff
@@ -0,0 +1,46 @@
...
+    public viwIssuesStepsService(viwIssuesStepsRepo vStepsRepo, IssuesStepsFactory iStepsFactory){
+        this.vStepsRepo = vStepsRepo;
+        this.iStepsFactory = iStepsFactory;
+    }
+
+    public viwIssuesSteps getIssueStepByID(Long id){
+        return vStepsRepo.findById(id).orElseThrow(() -> new RuntimeException(""));
+    }
+
+    public List<viwIssuesSteps> getAllIssueStepsByIssueId(){
```
Comment (ValdasPiekus) [inline src/main/java/com/kustudents/issuetracker/service/viwIssuesStepsService.java:28]: "Get all issue steps by issue id" - but I don't see an issue id variable anywhere In this function

File: src/main/java/com/kustudents/issuetracker/service/viwIssuesStepsService.java [hunk 17]
```diff
@@ -0,0 +1,46 @@
...
+    }
+
+    public viwIssuesSteps createIssueSteps(viwIssuesSteps vIssuesStep){
+        return vStepsRepo.save(vIssuesStep);
+    }
+
+    public void createFirstIssueStep(viwIssues vIssues){
+        //TODO: map actual login... somehow
+        vStepsRepo.save(iStepsFactory.createFirstIssueStep("root", vIssues, vIssues.getDescription()));
+        System.out.println(vIssues);
```
Comment (ValdasPiekus) [inline src/main/java/com/kustudents/issuetracker/service/viwIssuesStepsService.java:39]: As I undestand, you wanted to print out the issue to be able to see it yourself. If you want to use logging into the console. You can add "@Slf4j" to viwIssuesStepsService class, and then use it's logging capabilities. A simple log example: log.info("Getting academy by id: {}", academyId);

**Review by ValdasPiekus — APPROVED** [review]
